notebooks/working/_03-tb-deconfounder-predictive-checks.md

```diff
@@ -0,0 +1,683 @@
+    ]
+)
+
+# draw the causal model
+V_Drive_Alone.draw()
+```
+
+```python
+# Create a list of the variables in the drive alone utility
+drive_alone_variables =\
+    ['total_travel_distance',
+     'total_travel_cost',
+     'total_travel_time',
+     'num_cars',
+     'num_licensed_drivers'
+    ]
+
+# Create a sub-dataframe with those variables
+drive_alone_df =\
+    df.loc[df['mode_id'] == 1, drive_alone_variables]
+
+# Get the means and standard deviations of those variables
+drive_alone_means = drive_alone_df.mean()
+drive_alone_means.name = 'mean'
+
+drive_alone_stds = drive_alone_df.std()
+drive_alone_stds.name = 'std'
+
+# Look at the computed means and standard deviations
+print(pd.DataFrame([drive_alone_means, drive_alone_stds]))
+```
+
+## Specify the factor model that is to be checked
+
+In Wang and Blei's deconfounder technique, we fit a factor model to the variables in one's outcome model.
+
+The factor model being considered here is:
+
+$
+\begin{aligned}
+X_{\textrm{standardized}} &= Z * W + \epsilon\\
+\textrm{where } \epsilon &= \left[ \epsilon_1, \epsilon_2, ..., \epsilon_D \right]\\
+\epsilon_d &\in \mathbb{R}^{\textrm{N x 1}}\\
+\epsilon_d &\sim \mathcal{N} \left(0, \sigma \right) \forall d \in \left\lbrace 1, 2, ... D \right\rbrace\\
+Z &\in \mathbb{R}^{\textrm{N x 1}}\\
+Z &\sim \mathcal{N} \left(0, 1 \right)\\
+W &\in \mathbb{R}^{1 x D}\\
+W &\sim \mathcal{N} \left(0, 1 \right)\\
+N &= \textrm{Number of rows in X_standardized}\\
+D &= \textrm{Number of columns in X_standardized}
+\end{aligned}
+$
+
+```python
+# Note the number of dimensions
+num_dimensions = len(drive_alone_variables)
+
+# Specify the prior distributions for the factor
+# model of the standardized drive alone dataframe
+w_dist_prior = norm(loc=0, scale=1)
+z_dist_prior = norm(loc=0, scale=1)
+
+sigma_prior = 0.1
+epsilon_dists_prior =\
+    [norm(loc=0, scale=sigma_prior) for i in range(num_dimensions)]
+```
+
+## Generate prior predictive samples
+
+```python
+# Get the number of observations for this utility
+num_drive_alone_obs = drive_alone_df.shape[0]
+
+# Set a seed for reproducibility
+np.random.seed(721)
+
+# Get prior samples of X_standardized
+w_samples_prior =\
+    w_dist_prior.rvs((1, num_dimensions, NUM_PRIOR_SAMPLES))
+z_samples_prior =\
+    z_dist_prior.rvs((num_drive_alone_obs, 1, NUM_PRIOR_SAMPLES))
+
+epsilon_samples_prior =\
+    np.concatenate(
+        [dist.rvs((num_drive_alone_obs, NUM_PRIOR_SAMPLES))[:, None, :]
+         for dist in epsilon_dists_prior],
+        axis=1)
+
+x_standardized_samples_prior =\
+    (np.einsum('mnr,ndr->mdr', z_samples_prior, w_samples_prior) +
+     epsilon_samples_prior)
+
+# Get samples of X on the original scale of each variable
+x_samples_prior =\
+    (x_standardized_samples_prior *
+     drive_alone_stds[None, :, None] +
+     drive_alone_means[None, :, None])
+```
+
+```python
+# Look at the dimensions of the prior predictive samples
+print(x_samples_prior.shape)
```

Comment:
No comment here, just wanted to say this is maybe what the shape of my simulated data should be. Store everything in a 3 dimensional array? We can discuss later.

Reply:
For sure. It's a little harder for your case as the number of observations isn't constant, so numpy wouldn't let you store things in a 3d array without some hacks like setting some rows to be all null or something similar.